server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java

```diff
@@ -353,20 +368,52 @@ public void writeTo(StreamOutput out) throws IOException {
     }
 
     /**
-     * Sets the search query for this request.
-     *
-     * @see org.elasticsearch.index.query.QueryBuilders
+     * Sets the query for this request.
      */
-    public SearchSourceBuilder query(QueryBuilder query) {
-        this.queryBuilder = query;
+    public SearchSourceBuilder query(QueryBuilder queryBuilder) {
+        subSearchSourceBuilders = new ArrayList<>();
+        if (queryBuilder != null) {
```

Comment:
do we run into troubles if we enforce this to be non null? Was it used before to reset a previously set query perhaps? If so doesn't the new impl change that behaviour?

Reply:
This mimics the behavior prior by where an empty sub searches list is the same a null query.